File: app/src/main/java/ru/vtb/ppcd/example/processor/MessageProcessorFunction.java
package ru.vtb.ppcd.example.processor;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.apache.camel.Exchange;
import org.apache.camel.Processor;

import static ru.vtb.ppcd.example.utils.ExchangeUtils.toExchangeContext;
import static ru.vtb.ppcd.example.utils.ExchangeUtils.updateExchange;


public interface MessageProcessorFunction<T, R> extends Processor {

    default void process(Exchange exchange) {
        var exchangeContext = toExchangeContext(exchange, getObjectMapper(), processClass());

        R processedMessage = processMessage(exchangeContext);

        ExchangeContext<R> updatedExchangeContext = ExchangeContext.<R>builder()
                .headers(exchangeContext.getHeaders())
                .messageId(exchangeContext.getMessageId())
                .message(processedMessage)
                .build();

        updateExchange(exchange, getObjectMapper(), updatedExchangeContext);
    }

    R processMessage(ExchangeContext<T> exchangeContext);

    Class<T> processClass();

    ObjectMapper getObjectMapper();
}

File: ms-ppcd-example-app/src/main/kotlin/ru/vtb/ppcd/example/processor/MessageProcessorFunction.kt
package org.example.ru.vtb.ppcd.example.processor

import com.fasterxml.jackson.databind.ObjectMapper
import org.apache.camel.Exchange
import org.apache.camel.Processor
import org.example.ru.vtb.ppcd.example.model.ExchangeContext
import org.example.ru.vtb.ppcd.example.utils.ExchangeUtils.Companion.toExchangeContext
import org.example.ru.vtb.ppcd.example.utils.ExchangeUtils.Companion.updateExchange

interface MessageProcessorFunction<T, R> : Processor {
    override fun process(exchange: Exchange) : Unit =
        toExchangeContext(exchange, getObjectMapper(), processClass())
            .let { exchangeContext ->
                run {
                    ExchangeContext(
                        message = processMessage(exchangeContext),
                        messageId = exchangeContext.messageId,
                        headers = exchangeContext.headers
                    ).let { updatedContext ->
                        updateExchange(exchange, getObjectMapper(), updatedContext)
                    }
                }
            }

    fun processMessage(exchangeContext: ExchangeContext<T>): R

    fun processClass(): Class<T>

    fun getObjectMapper(): ObjectMapper
}